pkg/controller/machinepool/nutanixactuator.go

```diff
@@ -0,0 +1,229 @@
+// a list of MachineSets to sync to the remote cluster
+type NutanixActuator struct {
+	client        client.Client
+	masterMachine *machineapi.Machine
+}
+
+var _ Actuator = &NutanixActuator{}
+
+// NewNutanixActuator is the constructor for building a NutanixActuator
+func NewNutanixActuator(client client.Client, masterMachine *machineapi.Machine) (*NutanixActuator, error) {
+	actuator := &NutanixActuator{
+		client:        client,
+		masterMachine: masterMachine,
+	}
+	return actuator, nil
+}
+
+func decodeNutanixMachineProviderSpec(rawExtension *runtime.RawExtension) (*machinev1.NutanixMachineProviderConfig, error) {
+	if rawExtension == nil {
+		return &machinev1.NutanixMachineProviderConfig{}, nil
+	}
+
+	spec := new(machinev1.NutanixMachineProviderConfig)
+	if err := json.Unmarshal(rawExtension.Raw, &spec); err != nil {
+		return nil, fmt.Errorf("error unmarshalling providerSpec: %v", err)
+	}
+
+	return spec, nil
+}
+
+// getRHCOSImageNameFromMasterMachine retrieves the RHCOS image name from the master machine provider spec.
+// Note: The machine image name is reliably set only in IPI installations. In UPI installations, this field may not be populated.
+func getRHCOSImageNameFromMasterMachine(masterMachine *machineapi.Machine, cd *hivev1.ClusterDeployment, logger log.FieldLogger) (string, error) {
+	providerSpec, err := decodeNutanixMachineProviderSpec(masterMachine.Spec.ProviderSpec.Value)
+	if err != nil {
+		return "", errors.Wrap(err, "failed to decode provider spec while retrieving RHCOS image name from master machine")
+	}
+
+	if providerSpec.Image.Name != nil && *providerSpec.Image.Name != "" {
+		logger.Infof("using RHCOS image name from master machine providerSpec: %s", *providerSpec.Image.Name)
+		return *providerSpec.Image.Name, nil
+	}
+
+	return "", errors.Errorf("no RHCOS image name found in provider spec for ClusterDeployment %s/%s ", cd.Namespace, cd.Name)
+}
+
+// GenerateMachineSets satisfies the Actuator interface and will take a clusterDeployment and return a list of MachineSets
+// to sync to the remote cluster.
+func (a *NutanixActuator) GenerateMachineSets(cd *hivev1.ClusterDeployment, pool *hivev1.MachinePool, logger log.FieldLogger) ([]*machineapi.MachineSet, bool, error) {
+	if cd.Spec.ClusterMetadata == nil {
+		return nil, false, errors.New("ClusterDeployment does not have cluster metadata")
+	}
+
+	if cd.Spec.Platform.Nutanix == nil {
+		return nil, false, errors.New("ClusterDeployment is not for Nutanix")
+	}
+	if pool.Spec.Platform.Nutanix == nil {
+		return nil, false, errors.New("MachinePool is not for Nutanix")
+	}
+
+	computePool := baseMachinePool(pool)
+	dataDisks, err := a.getNutanixDataDisks(pool, logger)
+	if err != nil {
+		if updateErr := a.setUnsupportedConfigurationCondition(pool, logger, "InvalidDataDisk", "data source must specify a UUID"); updateErr != nil {
+			return nil, false, updateErr
+		}
+		return nil, false, nil
```

Comment:
nit, this could be pared down to
```suggestion
		return nil, false, a.setUnsupportedConfigurationCondition(pool, logger, "InvalidDataDisk", "data source must specify a UUID")
```

Similar below. But not really worth fixing.